### DISCUSSION TITLE
Bluetooth not working on nvdia jetson nx dev kit

### DISCUSSION THREAD

**Discussion by DhanushLingeswaran15** [OE4T/meta-tegra#853] [Q&A] [answered]
Hi ,

I am using the branch dunfell-l4t-r32.5.0 and could not get Bluetooth working.

tried the following
bluetoothctl  
Agent registered
[bluetooth]# list
[bluetooth]#  (empty)

hcitool dev
output:
Devices:
(empty)

hciconfig list
hci0:	Type: Primary  Bus: USB
	BD Address: 00:00:00:00:00:00  ACL MTU: 0:0  SCO MTU: 0:0
	DOWN 
	RX bytes:0 acl:0 sco:0 events:0 errors:0
	TX bytes:0 acl:0 sco:0 commands:0 errors:0

hciconfig hci0 up
Can't init device hci0: Input/output error (5)

Restarted the bluetoothd daemon. daemon running but no other change.

also out of curiosity tried calling /usr/sbin/tegra-bluetooth-init
and got the following error
port --use_baudrate_for_download could not be opened, error 2

Any ideas to get bluetooth up would be great.

**Comment by madisongh** [comment]
* Does `lsmod` show the `rtk_btusb` module is loaded? 
* Does `lsusb` list the BT radio?

If yes to both of those, what does `rfkill list` show? If it's soft blocked, try `rfkill unblock bluetooth`.
If `hciconfig list` still shows it as down, try `bluetoothctl power on`

The `tegra-bluetooth-init` script is only applicable on the Jetsons with the Broadcom Wi-Fi/Bluetooth module.

**Comment by DhanushLingeswaran15** [comment]
Hi, thanks for the reply.

Still no luck.Tried out the following,

root@dhanush:~# bluetoothctl power on
No default controller available    (I could not set a default controller because the list is empty). 

root@dhanush:~# hciconfig hci0 up
Can't init device hci0: Input/output error (5)

root@dhanush:~# lsusb
Bus 002 Device 002: ID 0bda:0489 Realtek Semiconductor Corp. 4-Port USB 3.0 Hub
Bus 002 Device 001: ID 1d6b:0003 Linux Foundation 3.0 root hub
Bus 001 Device 005: ID 13d3:3549 IMC Networks Bluetooth Radio
Bus 001 Device 004: ID 046a:00a1 Cherry GmbH SmartCard Reader Keyboard KC 1000 SC
Bus 001 Device 003: ID 046d:c040 Logitech, Inc. Corded Tilt-Wheel Mouse
Bus 001 Device 002: ID 0bda:5489 Realtek Semiconductor Corp. 4-Port USB 2.0 Hub
Bus 001 Device 001: ID 1d6b:0002 Linux Foundation 2.0 root hub

root@dhanush:~# lsmod
    Not tainted
bnep 19014 2 - Live 0xffffff8000fa5000
rtl8822ce 3453734 0 - Live 0xffffff800137c000
rtk_btusb 68665 0 - Live 0xffffff800134a000
btusb 44955 0 - Live 0xffffff8001328000
btrtl 7877 1 btusb, Live 0xffffff800131e000
btbcm 10207 1 btusb, Live 0xffffff8001310000
btintel 13507 1 btusb, Live 0xffffff8001300000
nvgpu 1726742 42 - Live 0xffffff80010de000
cfg80211 695716 1 rtl8822ce, Live 0xffffff8000fec000
mttcan 72799 0 - Live 0xffffff8000fd1000
can_dev 17217 1 mttcan, Live 0xffffff8000fc0000
spidev 14635 0 - Live 0xffffff8000fa0000

root@dhanush:~# rfkill list
0: hci0: bluetooth
	Soft blocked: no
	Hard blocked: no

Also, just to check the hardware i tried a sd card with the jetpack image. Bluetooth hardware seems to work fine.

**Comment by DhanushLingeswaran15** [comment, marked as answer]
I also found this in dmesg,

 7.619477] rtk_btusb: Realtek Bluetooth USB driver ver 3.1
[    7.619483] rtk_btcoex: rtk_btcoex_init: version: 1.2
[    7.619486] rtk_btcoex: create workqueue
[    7.633736] rtk_btcoex: alloc buffers 1792, 2432 for ev and l2
[    7.633830] rtk_btusb: btusb_probe intf->cur_altsetting->desc.bInterfaceNumber 0
[    7.633836] rtk_btusb: btusb_probe can_wakeup 1, may wakeup 0
[    7.633839] rtk_btusb: patch_add
[    7.633842] rtk_btusb: auto suspend is disabled
[    7.633846] rtk_btusb: pid = 0x3549
[    7.633850] rtk_btusb: patch_add: Reset gEVersion to 0xff
[    7.633859] rtk_btusb: set_bit(HCI_QUIRK_RESET_ON_CLOSE, &hdev->quirks);
[    7.644430] rtk_btusb: btusb_open start
[    7.644435] rtk_btusb: btusb_open hdev->promisc ==0
[    7.644438] rtk_btusb: download_patch start
[    7.644443] rtk_btusb: chip type value: 0x73
[    7.644554] rtk_btusb: HCI reset.
[    7.644604] rtk_btusb: btusb_probe: done
[    7.644658] usbcore: registered new interface driver rtk_btusb
[    7.655050] rtk_btusb: read_ver_rsp->lmp_subver = 0x8822
[    7.655056] rtk_btusb: read_ver_rsp->hci_rev = 0xc
[    7.655059] rtk_btusb: patch_entry->lmp_sub = 0x8822
[    7.655062] rtk_btusb: load_firmware start
[    7.655066] rtk_btusb: lmp_version = 0x8822
[    7.655069] rtk_btusb: config filename rtl8822cu_config
[    7.709720] usb 1-3: Direct firmware load for rtl8822cu_config failed with error -2
[    7.709889] usb 1-3: Falling back to user helper
[    7.711232] rtk_btusb: fw name is  rtl8822cu_fw
[    7.712821] usb 1-3: Direct firmware load for rtl8822cu_fw failed with error -2
[    7.712975] usb 1-3: Falling back to user helper
[    7.714008] rtk_btusb: load firmware failed!
[    7.714017] rtk_btusb: Rtk patch end -1
[    7.714025] rtk_btusb: btusb_open failed

Could this be the problem?

**Reply by madisongh** [reply]
Yep, if the firmware isn't present for loading, that would be a problem.  Did your image include the `tegra-firmware-rtl8822` package?

**Reply by timongentzsch** [reply]
had the same problem and could also solve it with manual addition of the `tegra-firmware-rtl8822` package. However, I was surprised that this is necessary, since this package is set in the `MACHINE_EXTRA_RRECOMMENDS`: https://github.com/OE4T/meta-tegra/blob/abf7ba32dba10d7476aea7a6d38334f5d561a33d/conf/machine/include/xavier-nx.inc#L19  According to the mega manual the packages in this variable are included in all images based on packagegroup-base. However, I could not reproduce it this way. Did i miss something?

**Reply by madisongh** [reply]
> According to the mega manual the packages in this variable are included in all images based on packagegroup-base. However, I could not reproduce it this way. Did i miss something?

Yeah, it should get included automatically for most images. If you're building a custom image recipe, make sure it's including `packagegroup-base` or at least `packagegroup-machine-base`.

**Comment by DhanushLingeswaran15** [comment]
Hey! 
That was the problem! Works fine now. 
Thanks a lot :+1:
